src/braintree-manager/braintree-manager.ts

```diff
@@ -0,0 +1,235 @@
+    });
+  }
+
+  /** @inheritdoc */
+  donationSuccessful(options: {
+    successResponse: SuccessResponse;
+    upsellSuccessResponse?: SuccessResponse;
+  }): void {
+    this.endpointManager.donationSuccessful(options);
+  }
+
+  private deviceDataCollectionStarted = false;
+
+  /**
+   * Collect Braintree device data. This is used to help fraud detection.
+   *
+   * @private
+   * @returns {Promise<void>}
+   * @memberof BraintreeManager
+   */
+  private async collectDeviceData(): Promise<void> {
+    if (this.deviceDataCollectionStarted) {
+      return;
+    }
+    this.deviceDataCollectionStarted = true;
+
+    const instance = await this.instance.get();
+    if (!instance) {
+      return;
+    }
+
+    return this.paymentClients.dataCollector
+      .get()
+      .then((collector?: braintree.DataCollector) => {
+        return collector?.create({ client: instance, kount: false, paypal: true });
+      })
+      .then(instance => {
+        this.deviceData = instance?.deviceData;
+      });
+  }
+
+  /**
+   * Braintree Authorization Token
+   *
+   * @private
+   * @type {string}
+   * @memberof BraintreeManager
+   */
+  private authorizationToken: string;
+
+  /**
+   * The endpoint manager for network communications
+   *
+   * @private
+   * @type {BraintreeEndpointManagerInterface}
+   * @memberof BraintreeManager
+   */
+  private endpointManager: BraintreeEndpointManagerInterface;
+
+  /**
+   * The payment clients container containing the braintree, paypal, and google clients
+   *
+   * @private
+   * @type {PaymentClientsInterface}
+   * @memberof BraintreeManager
+   */
+  private paymentClients: PaymentClientsInterface;
+
+  private hostingEnvironment: HostingEnvironment = HostingEnvironment.Development;
+
+  constructor(options: {
+    authorizationToken: string;
+    paymentClients: PaymentClientsInterface;
+    endpointManager: BraintreeEndpointManagerInterface;
+    hostedFieldConfig: HostedFieldConfiguration;
+    hostingEnvironment: HostingEnvironment;
+    venmoProfileId?: string;
+    googlePayMerchantId?: string;
+    referrer?: string;
+    loggedInUser?: string;
+  }) {
+    this.authorizationToken = options.authorizationToken;
+    this.endpointManager = options.endpointManager;
+    this.hostingEnvironment = options.hostingEnvironment;
+    this.paymentClients = options.paymentClients;
+
+    this.referrer = options.referrer;
+    this.loggedInUser = options.loggedInUser;
+
+    this.paymentProviders = new PaymentProviders({
+      braintreeManager: this,
+      paymentClients: this.paymentClients,
+      venmoProfileId: options.venmoProfileId,
+      googlePayMerchantId: options.googlePayMerchantId,
+      hostingEnvironment: options.hostingEnvironment,
+      hostedFieldConfig: options.hostedFieldConfig,
+    });
+
+    this.instance = new PromisedSingleton<braintree.Client>({
+      generator: this.paymentClients.braintreeClient.get().then((client: braintree.Client) => {
```

Comment:
will we ever need to retry or handle a failure here?

Reply:
This returns a Promise so anything requesting the instance can still handle it like a promise:
```
try {
  const client = await braintreeManager.instance.get();
} catch (err) {
  // do something error-y
}
```
As for retrying, the low-level thing that would fail here is the lazy loader that loads the `<script>` tag so I'm not sure there's much we can do if that fails.